File: dataStructure/二分查找/binary_search.ts
// 二分查找(双闭区间)
function binarySearch(nums: number[], target: number): number {
  // 初始化双闭区间[0, n-1]
  // 即i，j分别指向数组首元素，尾元素
  let i = 0,
      j = nums.length - 1;
  while(i <= j) {
    // 计算中点索引 m
    const m = Math.floor(i + (j - i) / 2);
    if(nums[m] < target) {
      i = m + 1;
    } else if(nums[m] > target) {
      j = m - 1;
    } else {
      return m;
    }
  }
  return -1;
}

// 二分查找（左闭右开区间）
function binarySearchLCRO(nums: number[], target: number): number {
  // 初始化左闭右开区间[0, n),即i，j分别指向数组首元素、尾元素+1
  let i = 0,
      j = nums.length;
  while(i < j) {
    const m = Math.floor(i + (j - i) / 2);
    if(nums[m] < target) {
      i = m + 1;
    } else if(nums[m] > target) {
      j = m;
    } else  {
      return m;
    }
  }
  return -1;
}

export { binarySearch, binarySearchLCRO }

File: dataStructure/二分查找/binarysearch.java
package dataStructure.二分查找;

public class binarysearch {

    // 二分查找递归写法
    // 这里也是无重复元素
    public int binarySearch(int[] array, int low, int high, int target) {
        // 递归终止条件
        if (low <= high) {
            int mid = low + ((high - low) >> 1);
            if (array[mid] == target) return mid;
            if (array[mid] > target) return binarySearch(array, low, mid-1, target);
            if (array[mid] < target) return binarySearch(array, mid+1, high, target);
        }
        return -1;
    }


    // 如果是有重复元素，如果查找的元素重复，则找左侧第1个
    // 这里的方法是线性向左查找
    public static int search(int[] nums, int target) {
        if (nums == null || nums.length == 0) return -1;
        int left = 0;
        int right = nums.length - 1;
        while (left <= right) {
            int mid = left + ((right - left) >> 1);
            if (nums[mid] < target) {
                left = mid + 1;
            } else if (nums[mid] > target) {
                right = mid - 1;
            } else {
                // 找到之后，往左找
                while (mid != 0 && nums[mid] == target) mid--;
                if (mid == 0 && nums[mid] == target) return mid;
                return mid + 1;
            }
        }
        return -1;
    }



}
